lib/tasks/plan_validation_reports.rake

```diff
@@ -0,0 +1,79 @@
+# To run all reports please use this rake command: RAILS_ENV=production bundle exec rake plan_validation:reports["2019-12-01"]
+
+namespace :plan_validation do
+
+  desc "reports generation after plan loading"
+  task :reports, [:active_date] => :environment do |_task, args|
+    puts "Reports generation started" unless Rails.env.test?
+    puts "Reports generation started for Report1" unless Rails.env.test?
+    Rake::Task['plan_validation:report1'].invoke(args[:active_date])
+    puts "Reports generation started for Report2" unless Rails.env.test?
+    Rake::Task['plan_validation:report2'].invoke(args[:active_date])
+  end
+
+  #To run first report: RAILS_ENV=production bundle exec rake plan_validation:report1["2020-01-01"]
+  desc "Details about Plan Count by Carrier, Coverage and Tier"
+  task :report1, [:active_date] => :environment do |_task, args|
+    active_date = args[:active_date].to_date
+    active_year = active_date.year
+    CSV.open("#{Rails.root}/plan_validation_report1_#{args[:active_date]}.csv", "w", force_quotes: true) do |csv|
+      csv << %w[PlanYearId CarrierId CarrierName PlanTypeCode Tier Count]
+      products = ::BenefitMarkets::Products::Product.by_year(args[:active_date])
+      products.all.each do |product|
+        active_year = product.active_year
+        carrier_id = product.hios_id[0..4]
+        carrier_name = product.issuer_profile.abbrev
+        plan_type_code = product.kind == :health ? "QHP" : "QDP"
+        tier = product.metal_level_kind
+        product_count = ::BenefitMarkets::Products::Product.by_year(active_year).where(hios_id: /#{carrier_id}/i, metal_level_kind: tier).count
+        csv << [active_year, carrier_id, carrier_name, plan_type_code, tier, product_count]
+      end
+      puts "Successfully generated Plan validation Report1"
+    end
+  end
+
+  #To run second report: RAILS_ENV=production bundle exec rake plan_validation:report2["2020-01-01"]
+  desc "Rating Area and Age Based Plan Rate Sum"
+  task :report2, [:active_date] => :environment do |_task, args|
+    active_date = args[:active_date].to_date
+    active_year = active_date.year
+    CSV.open("#{Rails.root}/plan_validation_report2_#{active_year}.csv", "w", force_quotes: true) do |csv|
+      csv << %w[PlanYearId CarrierId CarrierName RatingArea Age Sum]
+      issuer_hios_ids = BenefitSponsors::Organizations::ExemptOrganization.issuer_profiles.map(&:profiles).flatten.flat_map(&:issuer_hios_ids).map(&:to_i)
+
+      rating_area_ids = ::BenefitMarkets::Locations::RatingArea.where(active_year: active_year).inject({}) do |data, ra|
+        data[ra.id.to_s] = ra.exchange_provided_code
+        data
+      end
+
+      issuer_hios_ids.each do |issuer_hios_id|
+        products = ::BenefitMarkets::Products::Product.by_year(active_year).where(hios_id: /#{issuer_hios_id}/)
+        rating_area_ids.each do |rating_area_key, rating_area_value|
+          premium_tables = products.map(&:premium_tables).flatten.select do |prem_tab|
```

Comment:
Use this scope on the product. 
`effective_with_premiums_on` and pass the date to it.

Reply:
`products.effective_with_premiums_on(active_date)`

Reply:
Oh, nice I was not aware of this scope. Sure, I will use the scope thanks.

Reply:
wait, it returns the products. not the premium_tables. you can leave it as is.